calico_containers/pycalico/ipam.py

```diff
@@ -199,13 +208,61 @@ def _claim_block_affinity(self, host, block_cidr):
                 # no longer exists.
                 pass
 
-            # Throw a key error to let the caller know the block wasn't free
-            # after all.
+            # Throw a HostAffinityClaimedError to let the caller know the block
+            # wasn't free after all.
             raise HostAffinityClaimedError("Block %s already claimed by %s",
-                                           block_id, block.host_affinity)
+                                           block_cidr, block.host_affinity)
+
         # successfully created the block.  Done.
         return
 
+    def _release_block_affinity(self, host, block_cidr):
+        """
+        Release a block we think is owned by the specified host.
+
+        If there are no IPs assigned in the block then delete the block.  If
+        there are IPs assigned, remove affinity of the block from the host.
+
+        Raises HostAffinityNotClaimedError if the block is claimed by a
+        different host.
+        Raises KeyError if the block does not exist.
+        """
+        for _ in xrange(RETRIES):
+            block = self._read_block(block_cidr)
+            if block.host_affinity != host:
+                _log.info("Block host affinity is %s (expected %s) - not "
+                          "releasing", block.host_affinity, host)
+                raise HostAffinityClaimedByOtherError(
+                          "Block %s is claimed by %s",
+                          block_cidr, block.host_affinity)
+
+            try:
+                if block.is_empty():
+                    # The block is empty, so just delete the block.
+                    _log.debug("Block is empty - delete block")
+                    self._delete_block(block)
+                elif block.strict_host_affinity:
+                    # The block has strict host affinity, therefore it can
```

Comment:
Hm.

Is this right?  I'm not sure I fully understand the reasoning / use-case here (or I've just forgotten it).

If we've assigned a few IPs from this block (all on the same host) and then release affinity on this block, those endpoints may still have those IPs, and those IPs might be re-assigned to another host some time in the future.  Is this OK? We could end up with multiple endpoints with the same IP.

Feels like this branch should do the same as the one below, but perhaps I'm missing something.